aggregate.go

```diff
@@ -0,0 +1,430 @@
+	if err := as.pre.Collect(id, anns...); err != nil {
+		return err
+	}
+
+	// Consider eviction of old data.
+	if time.Since(as.lastEvicted) > as.MinEvictAge {
+		if err := as.evictBefore(time.Now().Add(-1 * as.MinEvictAge)); err != nil {
+			return err
+		}
+	}
+
+	// Grab the group for our span.
+	group, ok := as.group(id, anns...)
+	if !ok {
+		// We don't have a group for the trace, and can't create one (the
+		// spanName event isn't present yet).
+		return nil
+	}
+
+	// Unmarshal the events.
+	var events []Event
+	if err := UnmarshalEvents(anns, &events); err != nil {
+		return err
+	}
+
+	// Find the start and end time of the trace.
+	var (
+		eStart, eEnd time.Time
+		haveTimes    = false
+	)
+	for _, e := range events {
+		e, ok := e.(TimespanEvent)
+		if !ok {
+			continue
+		}
+		if !haveTimes {
+			haveTimes = true
+			eStart = e.Start()
+			eEnd = e.End()
+			continue
+		}
+		if v := e.Start(); v.UnixNano() < eStart.UnixNano() {
+			eStart = v
+		}
+		if v := e.End(); v.UnixNano() > eEnd.UnixNano() {
+			eEnd = v
+		}
+	}
+	if !haveTimes {
+		// We didn't find any timespan events at all, so we're done here.
+		return nil
+	}
+
+	// Update the group to consider this trace being one of the slowest.
+	group.update(eStart, eEnd, id.Trace, func(trace ID) {
+		// Delete the request trace from the output store.
+		if err := as.DeleteStore.Delete(trace); err != nil {
+			log.Printf("AggregateStore: failed to delete a trace: %s", err)
+		}
+	})
+
+	// Move traces from the limit store into the group, as needed.
+	for _, slowest := range group.Slowest {
+		// Find the trace in the limit store.
+		trace, err := as.pre.Trace(slowest.TraceID)
+		if err == ErrTraceNotFound {
+			continue
+		}
+		if err != nil {
+			return err
+		}
+
+		// Place into output store.
+		var walk func(t *Trace) error
+		walk = func(t *Trace) error {
+			err := as.DeleteStore.Collect(t.Span.ID, t.Span.Annotations...)
+			if err != nil {
+				return err
+			}
+			for _, sub := range t.Sub {
+				if err := walk(sub); err != nil {
+					return err
+				}
+			}
+			return nil
+		}
+		if err := walk(trace); err != nil {
+			return err
+		}
+
+		// Delete from the limit store.
+		err = as.pre.Delete(slowest.TraceID)
+		if err != nil {
+			return err
+		}
+	}
+
+	// As we're updating the aggregation event, we go ahead and delete the old
+	// one now.
+	if err := as.DeleteStore.Delete(group.Trace); err != nil {
+		return err
+	}
+
+	// Record an aggregate event with the given name.
+	rec := NewRecorder(SpanID{Trace: group.Trace}, as.DeleteStore)
+	rec.Name(group.Name)
+	ev := &AggregateEvent{
+		Name:  group.Name,
+		Times: group.Times,
+	}
+	for _, slowest := range group.Slowest {
+		if slowest.TraceID != 0 {
+			ev.Slowest = append(ev.Slowest, slowest.TraceID)
+		}
+	}
+	if as.Debug && len(ev.Slowest) == 0 {
+		log.Printf("AggregateStore: no slowest traces for group %q (consider increasing MaxRate)", group.Name)
+	}
+	rec.Event(ev)
+	if errs := rec.Errors(); len(errs) > 0 {
```

Comment:
@keegancsmith Very good point! When you commented I had actually just ran into the issue but hadn't yet figured out where it was, your comment led me right to it :smiley: 

I could reproduce it by hitting the server with ~10k GET requests using `httperf`:

```
httperf --server=127.0.0.1 --port=8699 --uri=/endpoint --rate=15 --num-conns=100000000000 --rate=100 --timeout=5 --hog
```

Which would render the Dashboard UI entirely unusable (_none_ would show up).

I've added a workaround by coupling `AggregateStore` with `MemoryStore` (so it can use it's lock for synchronization) -- I think it's highly unlikely anyone would use another type of `DeleteStore` in practice (because right now `MemoryStore` is the only `DeleteStore`), so it seems to be an okay solution for now.